Simple FOV

nightblade9 opened the pull request:
- Basic FOV calculation (distance between points)
- Expensive calculation is cached, not run each frame
- Monsters only attack if you're in their FOV
- Basic (hideous) fog-of-war for walls

nightblade9 committed:
Nerf enemy count so we can test FOV better

nightblade9 committed:
Basic FOV calculation (distance between points)
Works well enough, DisplaySystem needs the EventBus though

nightblade9 committed:
Monsters only attack if you're in their FOV
Un-nerf monster count

nightblade9 committed:
"Fog of war" blankets the dungeon

nightblade9 committed:
FOV calculation is cached from turn to turn

nightblade9 committed:
Super-basic FOV that "remembers" wall tiles

3akev commented on src/Scripts/Ecs/Core/Entity.gd:
```diff
@@ -4,6 +4,7 @@ const SpriteComponent = preload("res://Scripts/Ecs/Components/SpriteComponent.gd
 
 # type (as a string) => instance
 var _components = {}
+var sight_radius = 0
```
Shouldn't this be in a component? Walls don't have sight, do they?

nightblade9 replied:
It's true that walls don't have sight. Which component would sight go into? ... I guess `SightComponent` 👀

3akev commented on src/Scripts/Ecs/Systems/DisplaySystem.gd:
```diff
@@ -4,24 +4,52 @@ var _tiles_tilemap:TileMap
+			
+		return false
+		
+	return sqrt(pow(entity.position.x - _player.position.x, 2) + pow(entity.position.y - _player.position.y, 2)) <= _player.sight_radius
```
Use `entity.position.distance_to(_player.position)`. Built in methods are in C++, therefore probably faster, so we can get away with dropping that unsightly AABB too. Is it really necessary? Seems like premature optimization to me.

nightblade9 replied:
AABB is very fast and computationally cheap. It saves us from expensive distance calculations. You're right that `distance_to` is probably better - will do.

3akev commented on src/Scripts/Ecs/Systems/ChasePlayerSystem.gd:
```diff
@@ -18,11 +18,17 @@ func on_turn_end():
+		# Move only if player is in monster FOV.
+		# Expensive but accurate, assumes few monsters and done once per turn.
+		# If this is too expensive, switch to manhattan distance (no sqrt)
+		var distance = sqrt(pow(entity.position.x - _player.position.x, 2) + pow(entity.position.y - _player.position.y, 2))
```
Pass in `entity.position` and use `position.distance_to(_player.position)`.

3akev commented on src/Scripts/Ecs/Systems/DisplaySystem.gd:
```diff
@@ -4,24 +4,74 @@ var _tiles_tilemap:TileMap
+		var key = "%s, %s" % [entity.position.x, entity.position.y]
+		
+		# It's in our FOV, or it's a wall in fog-of-war
+		if  (_fov_cache.has(key) and _fov_cache[key] == true) or \
```
Can be simplified to `_fov_cache.get(key) == true`. `.get()` returns null if not found.

3akev commented on src/Scripts/Ecs/Systems/DisplaySystem.gd:
```diff
@@ -4,24 +4,74 @@ var _tiles_tilemap:TileMap
+			
+		return false
+		
+	return sqrt(pow(x - _player.position.x, 2) + pow(y - _player.position.y, 2)) <= _player.sight_radius
```
Ditto. Use built-in methods, please.

nightblade9 committed:
Code-review fixes from #12
Refactoring: `sight_radius` is a `SightComponent` field
Refactor: extract `PLAYER_SIGHT` constant

nightblade9 committed:
Code-review fixes for #12
Replace homebrew distance-between-points with (probably) C++ one that might be faster.
Simplify has_key and get_key
